tests/ui/unnecessary_lazy_eval.rs

```diff
@@ -321,3 +321,34 @@ fn panicky_arithmetic_ops(x: usize, y: isize) {
     let _x = false.then(|| f1 + f2);
     //~^ unnecessary_lazy_evaluations
 }
+
+mod issue14578 {
+    use std::future::Future;
+
+    pub struct OptionFuture<F> {
+        inner: Option<F>,
+    }
+
+    impl<F: Future> Future for OptionFuture<F> {
+        type Output = Option<F::Output>;
+
+        fn poll(self: std::pin::Pin<&mut Self>, cx: &mut std::task::Context<'_>) -> std::task::Poll<Self::Output> {
+            todo!()
+        }
+    }
+
+    impl<T> From<Option<T>> for OptionFuture<T> {
+        fn from(option: Option<T>) -> Self {
+            todo!()
+        }
+    }
+
+    async fn do_something() {
+        let random_bool = false;
+        let may_take_while: OptionFuture<_> = random_bool.then(async || take_while(42).await).into();
+    }
+
+    async fn take_while<T>(value: T) -> T {
+        todo!()
+    }
+}
```

Comment:
Wouldn't the following exhibit the same fix in a most compact way?

```suggestion
fn issue14578() {
    let _: Box<dyn Future<Output = i32>> = Box::new(true.then(async || 42).unwrap());
}
```